contracts/cw20-base/src/allowances.rs

```diff
@@ -0,0 +1,332 @@
+        amount: Uint128,
+    ) -> Result<Response, Self::Error> {
+        let (deps, env, info) = ctx;
+
+        let rcpt_addr = deps.api.addr_validate(&recipient)?;
+        let owner_addr = deps.api.addr_validate(&owner)?;
+
+        // Avoid doing state update in case of self to self transfer
+        if rcpt_addr == owner_addr {
+            return Ok(Response::new().add_attributes(vec![
+                attr("action", "transfer_from"),
+                attr("from", owner),
+                attr("to", recipient),
+                attr("by", info.sender),
+                attr("amount", amount),
+            ]));
+        }
+
+        if info.sender != owner {
+            // deduct allowance before doing anything else have enough allowance
+            self.deduct_allowance(deps.storage, &owner_addr, &info.sender, &env.block, amount)?;
+        }
+
+        self.balances.update(
+            deps.storage,
+            &owner_addr,
+            |balance: Option<Uint128>| -> StdResult<_> {
+                Ok(balance.unwrap_or_default().checked_sub(amount)?)
+            },
+        )?;
+        self.balances.update(
+            deps.storage,
+            &rcpt_addr,
+            |balance: Option<Uint128>| -> StdResult<_> { Ok(balance.unwrap_or_default() + amount) },
+        )?;
+
+        let res = Response::new()
+            .add_attribute("action", "transfer_from")
+            .add_attribute("from", owner)
+            .add_attribute("to", recipient)
+            .add_attribute("by", info.sender)
+            .add_attribute("amount", amount);
+        Ok(res)
+    }
+
+    /// Sends amount tokens from owner -> contract
+    /// if `env.sender` has sufficient pre-approval.
+    fn send_from(
+        &self,
+        ctx: (DepsMut, Env, MessageInfo),
+        owner: String,
+        contract: String,
+        amount: Uint128,
+        msg: Binary,
+    ) -> Result<Response, Self::Error> {
+        let (deps, env, info) = ctx;
+
+        let rcpt_addr = deps.api.addr_validate(&contract)?;
+        let owner_addr = deps.api.addr_validate(&owner)?;
+
+        // deduct allowance before doing anything else have enough allowance
+        self.deduct_allowance(deps.storage, &owner_addr, &info.sender, &env.block, amount)?;
+
+        // move the tokens to the contract
+        self.balances.update(
+            deps.storage,
+            &owner_addr,
+            |balance: Option<Uint128>| -> StdResult<_> {
+                Ok(balance.unwrap_or_default().checked_sub(amount)?)
+            },
+        )?;
+        self.balances.update(
+            deps.storage,
+            &rcpt_addr,
+            |balance: Option<Uint128>| -> StdResult<_> { Ok(balance.unwrap_or_default() + amount) },
+        )?;
+
+        let attrs = vec![
+            attr("action", "send_from"),
+            attr("from", &owner),
+            attr("to", &contract),
+            attr("by", &info.sender),
+            attr("amount", amount),
+        ];
```

Comment:
Still inconsistent attributes handling. And even if you have reason to do so this way (I assume you just want to create sense attribute as you will consume sender right after - I am fine with it), then at least remove `vec!` as it can be a normal array.